Enhance input decorator padding logic for character counter in text f…

RootHex200 opened the pull request:
Updated the padding calculations in the _RenderDecoration class to conditionally apply padding based on the presence of a character counter. This change ensures that the layout accommodates the counter correctly, improving the overall appearance and usability of text input fields.

Fix: #175591

before:
<img width="662" height="622" alt="image" src="https://github.com/user-attachments/assets/efd01c29-4aed-48c9-b719-1bbd84f2d406" />
after:
<img width="624" height="630" alt="image" src="https://github.com/user-attachments/assets/06cb2ea8-86cc-4200-aae0-1e864493c27e" />


*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

-  [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
-  [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
-  [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
-  [x] I signed the [CLA].
-  [x] I listed at least one issue that this PR fixes in the description above.
-  [x] I updated/added relevant documentation (doc comments with `///`).
-  [x] I added new tests to check the change I am making, or this PR is [test-exempt].
-  [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
-  [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

bleroux reviewed (commented):
@RootHex200 Thanks for working on this :pray: 

See my comment about creating a constant.
This PR will require a test.
The CI step 'Linux analyze' is failing due to code format. Using the dart auto formater should fix this issue.

victorsanni commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -922,8 +922,11 @@ class _RenderDecoration extends RenderBox
     };
 
+    // Only add padding when counter is present (maxLength is used)
+    final double counterPadding = counter != null ? 17 : 0.0;
```
Bumping on https://github.com/flutter/flutter/pull/175706/files#r2366936696 for this as well.

victorsanni reviewed (commented):
Hi @RootHex200, thanks for the PR. This PR will require a test in `flutter/test/material/input_decorator_test.dart` to land. Consider the other tests in the same file as a source of inspiration.

RootHex200 committed:
Enhance input decorator padding logic for character counter in text fields

Updated the padding calculations in the _RenderDecoration class to conditionally apply padding based on the presence of a character counter. This change ensures that the layout accommodates the counter correctly, improving the overall appearance and usability of text input fields.

bleroux reviewed (commented):
@RootHex200 Great update!

See my comments for test improvements.

One suggestion, once the review started it is better not to force push as it makes things more difficult to track for reviewers (before filing the PR, it is perfectly fine as the history of commits which led to the PR is usually more confusing than helpful).
I mention this because I remember seeing other changes in input_decorator.dart in your first iteration. I wanted to check if they were removed on purpose but I don't remember what lines were impacted).

RootHex200 committed:
Input decorator test improvement for character counter spacing

RootHex200 committed:
dart format fix

bleroux reviewed (commented):
Format is ok now, great!

See my comments about a change which will get the exact 16.0 result from the spec.
Let me know if it makes sense.

RootHex200 committed:
Refine input decorator padding logic for character counter

RootHex200 committed:
Fix typo in input decorator test comment for character counter spacing

bleroux reviewed (approved):
LGTM! Thanks for this useful fix :pray: 

Next steps after my approval are:
- the Google testing Ci check will run, let see if it fails (this can happen if some internal Google tests relied on the previous behavior).
- get a second review, apply requested changes and get approval. cc @victorsanni, if you get a chance to do that second review.
- once everything will be green Victor or I will add the autosubmit label and your PR will land some time later. :rocket:

victorsanni commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -922,8 +926,10 @@ class _RenderDecoration extends RenderBox
     };
 
+    // Only add padding when counter is present (maxLength is used).
+    final double counterPadding = counter != null ? _kSubtextCounterPadding : 0.0;
```
Does `counter` being non-null necessarily mean that a character count is visually present?

victorsanni replied:
Yes, they all had long helper/error text. Thanks for finding this.

bleroux replied:
I filed https://github.com/flutter/flutter/issues/175993 which describes the padding issue.

@RootHex200 Sorry for the inconvenience. My suggestion is to proceed as I described in my previous comment, it means updating the PR in the following ways:
- remove the `decoration.inputGap * 2` adjustement.
- update the test to restore your previous use of greaterThanOrEqualTo(8). Above the line with that expect clause add a comment explaining that the expected value should be 16 but it's not because of https://github.com/flutter/flutter/issues/175993

victorsanni commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -1265,7 +1271,12 @@ class _RenderDecoration extends RenderBox
 
-    final double helperErrorAvailableWidth = math.max(width - counterWidth, 0.0);
+    // Only add padding when counter is present (maxLength is used).
+    final double counterPadding = counter != null ? _kSubtextCounterPadding : 0.0;
```
Same here.

RootHex200 committed:
Refine input decorator padding calculations for character counter

bleroux reviewed (commented):
Thanks for the update! Just one last change and it should be fine.
See my comments.

RootHex200 committed:
Refine padding calculations in input decorator for character counter and update test comments for clarity

RootHex200 commented:
> Thanks for the update! Just one last change and it should be fine. See my comments.

done

bleroux reviewed (approved):
LGTM!
Let see now if Google testing step run again (sometimes it does not and Victor might have to trigger it) and if there are failures :crossed_fingers: .

QuncCccccc reviewed (approved):
LGTM. Thanks for the fix:)

I'm asking for approvals for cl/811419565. Once it's approved, we are good to go!

RootHex200 commented:
> LGTM. Thanks for the fix:)
> 
> I'm asking for approvals for cl/811419565. Once it's approved, we are good to go!

@QuncCccccc what about the approval

Thanks  ,
